Makefile

```diff
@@ -9,24 +9,52 @@ DEBIANOS := $(shell cat /etc/debian_version 2> /dev/null)
 TARGET_DIR := target
 BIN_NAME := attestation-agent
 
+SOURCE_ARCH := $(shell uname -m)
+
+ARCH ?= $(shell uname -m)
 DEBUG ?=
-LIBC ?=
+LIBC ?= gnu
 KBC ?=
 DESTDIR ?= $(PREFIX)/bin
+RUSTFLAGS_ARGS ?=
+
+PROTOC_S390X_VERSION := 21.1
+PROTOC_S390X_ARCHIVE := protoc-$(PROTOC_S390X_VERSION)-linux-s390_64.zip
 
 ifdef KBC
     feature := --no-default-features --features
 endif
 
 ifeq ($(LIBC), musl)
-    MUSL_ADD := $(shell rustup target add x86_64-unknown-linux-musl)
+    ifeq ($(ARCH), s390x)
+        $(error ERROR: Attestation agent does not support building with the musl libc target for s390x architecture!)
+    endif
+    MUSL_ADD := $(shell rustup target add ${ARCH}-unknown-linux-musl)
     ifneq ($(DEBIANOS),)
         MUSL_INSTALL := $(shell sudo apt-get install -y musl-tools) 
     endif
-    LIBC_FLAG := --target x86_64-unknown-linux-musl
-    TARGET_DIR := $(TARGET_DIR)/x86_64-unknown-linux-musl
 endif
 
+ifneq ($(SOURCE_ARCH), $(ARCH))
+    ifneq ($(DEBIANOS),)
+        GCC_COMPILER_PACKAGE_FOR_TARGET_ARCH := gcc-$(ARCH)-linux-$(LIBC)
+        GCC_COMPILER_FOR_TARGET_ARCH := $(ARCH)-linux-$(LIBC)-gcc
+        RUSTC_TARGET_FOR_TARGET_ARCH := $(ARCH)-unknown-linux-$(LIBC)
+        GCC_FOR_TARGET_ARCH_INSTALL := $(shell sudo apt-get install -y ${GCC_COMPILER_PACKAGE_FOR_TARGET_ARCH})
+        RUST_TARGET_FOR_TARGET_ARCH_INSTALL := $(shell rustup target add ${RUSTC_TARGET_FOR_TARGET_ARCH})
+        RUSTFLAGS_ARGS += -C linker=$(GCC_COMPILER_FOR_TARGET_ARCH)
+    else
+        $(error ERROR: Cross-compiling is only tested on Debian-like OSes)
+    endif
+endif
+
+ifeq ($(SOURCE_ARCH), s390x)
+    PROTOC_BINARY_INSTALL := $(shell wget https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_S390X_VERSION}/${PROTOC_S390X_ARCHIVE} && unzip -u ${PROTOC_S390X_ARCHIVE} && sudo cp bin/protoc /usr/local/bin/ && rm -f ${PROTOC_S390X_ARCHIVE})
+endif
```

Comment:
Isn't this some kind of prerequisite?
How do we know that protoc is installed for in x86_64 machines?
Maybe this should be outside the Makefile?
(similar question for `apt-get install ...` stuff we do here)

Reply:
Installing the separate `protoc` binary is required to compile attestation agent on s390x.

On x86_64 machines the attestation agent compiles without needing `protoc`.

I'm not sure if it's better to put the install for the protoc dependency and the `apt-get install ...` for the gcc cross-compiler toolchain inside the Makefile or outside of it. When building for attestation agent for x86_64 with the musl toolchain the installation of the `musl-tools` package is done inside the Makefile.

I don't maintain the build with musl but maybe that the musl-tools package is installed inside the Makefile means there's a precedent for putting installation of target-specific dependencies inside the Makefile.

Reply:
> I don't maintain the build with musl but maybe that the musl-tools package is installed inside the Makefile means there's a precedent for putting installation of target-specific dependencies inside the Makefile.

Yes, you're right. OK.